### PR TITLE
Fix breaking change in Searchable object type

### PR DESCRIPTION
**Pull request by Tofandel**
I believe this to be a Typescript bug though

### CODE REVIEW

**Comment by djfhe** [comment]
I will look into this tomorrow again, but this is not a bug in typescript :)

`Record<string | number, unknown>` describes a type having `string` and `number` as a index signature. But not everything has a signature in typescript. `interfaces` for example.
So:

```ts
interface TestInterface {
  prop1: number
}

declare const test: TestInterface

test satisfies Record<string | number, unknown> // fails, since test has no index signature
```

fails.

The `object` type describes any object which is also satisfied by functions and arrays.
The `Record<never, never>` type describes a empty object which is also satisfied by any function or array.

`object` and `Record<never, never>` constrained-wise are kinda the same in typescript:

```ts
// result is 'yes'
type AreObjectAndRecordNeverNeverTheSame = object extends Record<never, never>
	? Record<never, never> extends object
		? 'yes'
		: 'no'
	: 'no'
```
And since typescript is a structural typed language, we work with constrains and not the exact type most of the time :)

I think i just used `Record<never, never>` since tslint didn't like `object`.

`object` is the correct fix here or we can even revert back to `Record<never, never>`.
(I think) u can also remove the union, e.g. change:

```ts
type SearchableObject = object | unknown[]
```

to

```ts
type SearchableObject = object
```

since arrays are also just objects in typescript.

Here is a [ts playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgCoQM5gJLmvJZAbwChlkAHKAewujAE8BGALmRAFcBbAI2jMo06URgCY2WKKADmAqrXoMAzBIa9qAGxIBfEiUZ00mMAHkeAKwgIwyALzIASlepQAJgB4ACkMUBpCAwANMgcIADWINQA7iAAfHoA9ABUSWRJyNjIwNKRUCihEdEgANoAulkgyGAAFsAYyAhwGBDBGKCEwDau1Jjs1DZccGCQUMgwLsjUHKOQWA1NKNXQEAB0aQn6DIYAyhBwUAjVcDwaEGaW1gDC1CCScKAQrgDqndXnVmAvNbiuEAAe22yICG0xQ9icCBcHkkMmQAB92Nw+FBggVIjF4gYULt9odjqd3lcbncHs9XgBBEAMQk2ezUCwfTY7PYHI4nM4Mom3MBQe4gR5faoAUS4FEYNLsjmcbnc-IAbtBgvLoPESDBQtZgDdkM1cdUAEIMTxDaqCmmCn7-QE5EF5LzIf6QECueo41n4jkXMDXbm80lmzmfV6WgFA20QWIACgobE8AEpiMg8mBppUKLp1SBNdrdazDcaagGvYLKdTA-bHRBna6WXj2TSfST+WSaqWaVGY8h44nk6nKBmNWAtZVc4d8yaix9BSKxWWvRW-k6Xcg3XWCYHGzy+QLXjPxYGO7GE0QkxAU1A07oSL8EBp9ihIdyHX84KLTmx0FhcCMCCgEZ-TEDRIEnmEBIhsRAkDFCof0QTBWnaFAagCZBuj6GwjgVZA4AqX4-h1MNzwgEhRwNI0J1ec1g2dK1CNBSN-lfChTjjYDsIQKCwHqOAqUmQNgh4DgII0DBqDGQdh3qNosyQpYGGwvI+K9epQGQcw4DlJoECkMUSNrMiC1NSjAxLKkaQYl83wgVjSPHQtjOLCkzMDSNIwTWxYgaYlNFWDRqGkSMAHJxmoQK4xs-S7KMmoqNbZyvUjMoEzYyCIDFeoIAVKAGBqWEoiOGwMCGOoYGAXocIgUVGCUxlbPI+yYpM3cqrnD4LKYli9L1KLJ2sacWvMty7E8x9RNOFY-IC4LqFC8LkC6vN6uit4mpqPdWusRLSlYkgEhAvIMA4DQbDqZBAoYTBAqZFByTyGlKVcCEoQAOUy6BXqy1Alm2V8wRq6xnyXeonplZUUXYN6oHiABIAB+KVIVByGlUhzzK2rf6wBIaG4bOi6MCunG2ECyJCeJ0mgA) i used to compare different versions.

As i said, i will look  tomorrow a bit more into this and maybe even finish my work on https://github.com/clickbar/dot-diver/pull/11 so the search depth is not as anyoing and resource hungry as it currently is.

I hope this clarifies some things, if not feel free to ask.

P.S.: 
I would prefer `object` over `Record<never, never>` since it is a clearer outward facing interface than Record<never, never>.

**Comment by djfhe** [comment]
Okey ye, just `object` should be fine.

After looking a bit more into why no tests are failing with this issue, i found that we don't explicitly test the parameter type of `getByPath`/`setByPath`. we test its functionality in `index.test.ts` and test our internal types in `index.test-d.ts`, but never test the accepted types of theses functions exhaustively.

Maybe we should start doing this, what do you think @saibotk ?

**Comment by saibotk** [comment]
Yes we should do that.

So would object be incorrect then, if we started testing those types more?

**Comment by djfhe** [comment]
we may have to relax restriction in the checks of getByPath,  since we always check for an `typeof current !== 'object'` in the reduce, but functions are also a valid parameter for getByPath, since this is valid typescript:

```ts
const b: (() => void) & { foo?: string } = () => console.log('bar')

b['foo'] = 'bar'

console.log(b['foo'])
```

**Comment by djfhe** [comment]
sadly `object` since it accepts nearly every object, also `Number` or even intersections between primitive numbers and objects.

```ts
const foo: Number & { foo?: string } = 5
const bar: number & { foo?: string } = 5

type isPrimitiveNumberObject = number extends object ? true : false // false
type isNumberObject = Number extends object ? true : false // true

type isFooObject = typeof foo extends object ? true : false // true
type isBarObject = typeof bar extends object ? true : false // true
```

i think we should have the constrained for the given object here more relaxed, since it would required a more complicated type to check for this stuff and a more simple outward facing interface is more important here.

This should probably be handled by `PathValue` or the `Path`, rather than the constrained, since it is more clearer, if we just return no valid paths for `number & { foo?: string }` rather than a big error message, containing a 50 line type, for that parameter.

We maybe already check for this, but i don't think we have any tests covering this case.

**Review by saibotk — APPROVED** [review]
@djfhe do you approve too then?

File: src/index.ts [hunk 1]
```diff
@@ -216,8 +216,8 @@ type PathValueEntry<T, P extends PathEntry<T, Depth>, Depth extends number = 10>
   BuildTuple<Depth>
 >
 
-type SafeObject = Record<string | number | symbol, unknown>
-type SearchableObject = SafeObject | unknown[]
+type SafeObject = Record<string, unknown>
+type SearchableObject = object | unknown[]
```
Comment (djfhe) [inline src/index.ts]:
```suggestion
type SearchableObject = object
```

unknown[] is here unnecessary

**Review by djfhe — CHANGES_REQUESTED** [review]
Just wanted to change this

**Review by djfhe — APPROVED** [review]
Thanks for the fix! 👍

**Pull request merged**